be/src/exprs/string_functions.cpp

```diff
@@ -1049,6 +1050,384 @@ StatusOr<ColumnPtr> StringFunctions::repeat(FunctionContext* context, const Colu
+    std::vector<EncodedUtf8Char> encoded_from_values;
+    encode_utf8_chars(from_str, &encoded_from_values);
+
+    std::vector<EncodedUtf8Char> encoded_to_values;
+    encode_utf8_chars(to_str, &encoded_to_values);
+
+    const bool is_ascii_map = encoded_from_values.size() == from_str.size && encoded_to_values.size() == to_str.size;
+    dst_state->is_ascii_map = is_ascii_map;
+    if (is_ascii_map) {
+        size_t common_size = std::min(to_str.size, from_str.size);
+        int i = 0;
+        for (; i < common_size; i++) {
+            dst_state->ascii_map.emplace(from_str[i], to_str[i]);
+        }
+        for (; i < from_str.size; i++) {
+            dst_state->ascii_map.emplace(from_str[i], EncodedUtf8Char::ABSENT);
+        }
+    } else {
+        size_t common_size = std::min(encoded_to_values.size(), encoded_from_values.size());
+        int i = 0;
+        for (; i < common_size; i++) {
+            dst_state->utf8_map.emplace(encoded_from_values[i], encoded_to_values[i]);
+        }
+        for (; i < encoded_from_values.size(); i++) {
+            dst_state->utf8_map.emplace(encoded_from_values[i], EncodedUtf8Char{});
+        }
+    }
+}
+
+/**
+ * Translate a string by the UTF-8 map.
+ *
+ * @param src_encoded_values the string to translate, which has already been encoded.
+ * @param utf8_map the UTF-8 map.
+ * @param dst the destination to store the translated chars. The caller must guarantee there is enough room.
+ * @return [is_null, num_bytes].
+ * - `is_null` will be true, if the number of translated chars exceeds `OLAP_STRING_MAX_LENGTH`.
+ * - `num_bytes`: the number of translated chars.
+ */
+static inline std::pair<bool, size_t> translate_string_with_utf8_map(
+        const std::vector<EncodedUtf8Char>& src_encoded_values, const TranslateState::UTF8_MAP& utf8_map,
+        uint8_t* dst) {
+    size_t num_bytes = 0;
+    Slice dst_utf_char;
+    for (const auto& src_encoded_value : src_encoded_values) {
+        auto it = utf8_map.find(src_encoded_value);
+        if (it == utf8_map.end()) {
+            dst_utf_char = Slice(src_encoded_value);
+        } else if (!it->second.is_empty()) {
+            dst_utf_char = Slice(it->second);
+        } else {
+            continue;
+        }
+
+        if (num_bytes + dst_utf_char.size > OLAP_STRING_MAX_LENGTH) {
+            return {true, 0};
+        }
+        strings::memcpy_inlined(dst + num_bytes, dst_utf_char.data, dst_utf_char.size);
+        num_bytes += dst_utf_char.size;
+    }
+
+    return {false, num_bytes};
+}
+
+/**
+ * Translate a string by the ASCII map.
+ *
+ * @param src the string to translate.
+ * @param ascii_map the ASCII map.
+ * @param dst the destination to store the translated chars. The caller must guarantee there is enough room.
+ * @return the number of translated chars.
+ */
+static inline size_t translate_string_with_ascii_map(const Slice& src, const TranslateState::ASCII_MAP& ascii_map,
```

Comment:
it seems that we can use byte array instead of ASCII_MAP.

Reply:
Great!